mitreattack/navlayers/generators/usage_generator.py

```diff
@@ -0,0 +1,199 @@
+from stix2 import Filter
+from itertools import chain
+try:
+    from ..exporters.matrix_gen import MatrixGen
+    from ..core.exceptions import BadInput, typeChecker, categoryChecker
+    from ..core.layer import Layer
+except ValueError:
+    from mitreattack.navlayers.exporters.matrix_gen import MatrixGen
+    from mitreattack.navlayers.core.exceptions import BadInput, typeChecker, categoryChecker
+    from mitreattack.navlayers.core.layer import Layer
+except ImportError:
+    from navlayers.exporters.matrix_gen import MatrixGen
+    from navlayers.core.exceptions import BadInput, typeChecker, categoryChecker
+    from navlayers.core.layer import Layer
+
+
+class UnableToFindGroup(Exception):
+    pass
+
+
+class UnableToFindSoftware(Exception):
+    pass
+
+
+class UnableToFindMitigation(Exception):
+    pass
+
+
+class UsageGenerator:
+
+    def __init__(self, source, matrix='enterprise', local=None):
+        self.matrix_handle = MatrixGen(source, local)
+        self.domain = matrix
+        try:
+            self.source_handle = self.matrix_handle.collections[matrix]
+        except KeyError:
+            print(f"[UsageGenerator] - unable to load collection {matrix} (current source = {source}).")
+            raise BadInput
+
+    def remove_revoked(self, listing):
+        """
+        Remove revoked elements from the listing
+        :param listing: input element list
+        :return: input element list - revoked elements
+        """
+        removed = []
+        for x in listing:
+            if 'revoked' in x:
+                if x['revoked']:
+                    removed.append(x)
+        return [a for a in listing if a not in removed]
+
+    def get_group(self, match):
+        """
+        Retrieve group for a given match
+        :param match: matching pattern
+        :return: first matching group or UnableToFindGroup exception
+        """
+        a = []
+        if match.startswith('G'):
+            filts = [Filter('type', '=', 'intrusion-set')]
+            temp = self.source_handle.query(filts)
+            for b in temp:
+                for c in b['external_references']:
+                    if c['source_name'] == "mitre-attack" and c['external_id'] == match:
+                        return a
+        else:
+            filts = [Filter('type', '=', 'intrusion-set'),
+                     Filter('aliases', '=', match)]
```

Comment:
Unfortunately, I suspect matching based on name is always going to be less resilient than desired, but I can definitely add 'name' to one of the fields it can check.